### PR TITLE
Add basic matrix functionality

### PR DESCRIPTION
**Pull request by quintant**

### CODE REVIEW

File: permuta/misc/math.py [hunk 1]
```diff
@@ -10,3 +14,158 @@ def is_prime(n: int) -> bool:
...
+    def __setitem__(self, pos, value) -> None:
+        """Sets position in the matrix to val"""
+        assert 0 <= pos[0] < len(self)
+        assert 0 <= pos[1] < len(self)
+        self.elements[pos] = value
+
+    def __add__(self, other: "Matrix") -> "Matrix":
+        """Adds two matrices together. Both classes have to be binary or not binary"""
+        if not isinstance(other, Matrix):
+            return NotImplemented
```
Comment (JonSteinn) [inline permuta/misc/math.py:56]: `raise NotImplementedError("some msg")` ?

File: permuta/misc/math.py [hunk 2]
```diff
@@ -10,3 +14,158 @@ def is_prime(n: int) -> bool:
...
+        if self.binary != other.binary:
+            return NotImplemented
```
Comment (JonSteinn) [inline permuta/misc/math.py:58]: ...

File: permuta/misc/math.py [hunk 3]
```diff
@@ -10,3 +14,158 @@ def is_prime(n: int) -> bool:
...
+        binary: bool = True,
+    ) -> None:
+        if elements:
+            self.elements: Dict[Tuple[int, int], int] = {
+                key: (1 if binary and val else val) for key, val in elements.items()
+            }
+        else:
+            self.elements = dict()
+        self.binary = binary
+        self.__size = size
```
Comment (JonSteinn) [inline permuta/misc/math.py:41]: Single underscore?

File: permuta/misc/math.py [hunk 4]
```diff
@@ -10,3 +14,158 @@ def is_prime(n: int) -> bool:
...
+            for idx, val in chain(*funcs):
+                new_mat[idx] ^= val
+        else:
+            for idx, val in chain(*funcs):
+                new_mat[idx] += val
+        return new_mat
+
+    def __mul__(self, other: "Matrix") -> "Matrix":
+        if not isinstance(other, Matrix):
+            return NotImplemented
```
Comment (JonSteinn) [inline permuta/misc/math.py:72]: ...

File: permuta/misc/math.py [hunk 5]
```diff
@@ -10,3 +14,158 @@ def is_prime(n: int) -> bool:
...
+        if self.binary != other.binary:
+            return NotImplemented
```
Comment (JonSteinn) [inline permuta/misc/math.py:74]: ...

File: permuta/misc/math.py [hunk 6]
```diff
@@ -10,3 +14,158 @@ def is_prime(n: int) -> bool:
...
+            string += "|\n"
+        return string
+
+    def __repr__(self) -> str:
+        return f"Matrix({len(self)}, {self.elements})"
+
+    def __eq__(self, other: object) -> bool:
+        """Returns True if both matrices are equal."""
+        if not isinstance(other, Matrix):
+            return NotImplemented
```
Comment (JonSteinn) [inline permuta/misc/math.py:101]: ...

File: permuta/misc/math.py [hunk 7]
```diff
@@ -10,3 +14,158 @@ def is_prime(n: int) -> bool:
...
+        if self.binary != other.binary:
+            return NotImplemented
```
Comment (JonSteinn) [inline permuta/misc/math.py:103]: ...

File: permuta/patterns/perm.py [hunk 8]
```diff
@@ -87,6 +87,19 @@ def to_standard(cls, iterable: Iterable) -> "Perm":
     standardize = to_standard
     from_iterable = to_standard
 
+    @classmethod
+    def from_matrix(cls, matrix: "Matrix") -> "Perm":
+        """Returns the perm corresponding to the matrix given."""
+        if matrix.test_sum() != len(matrix):
+            raise Exception("Invalid matrix used")
```
Comment (JonSteinn) [inline permuta/patterns/perm.py:94]: Pick a more precise exception type

File: permuta/patterns/perm.py [hunk 9]
```diff
@@ -87,6 +87,19 @@ def to_standard(cls, iterable: Iterable) -> "Perm":
...
+        perm_list: List[Optional[int]] = [None for _ in range(len(matrix))]
+        for (row, col), val in matrix.elements.items():
+            if val:
+                perm_list[row] = col
+        if None in perm_list:
+            raise Exception("Invalid matrix used")
```
Comment (JonSteinn) [inline permuta/patterns/perm.py:100]: ...

File: permuta/__init__.py [hunk 10]
```diff
@@ -12,4 +13,6 @@
     "BivincularPatt",
     "CovincularPatt",
     "VincularPatt",
+    "Matrix",
+    "LoTriMatrix",
```
Comment (enadeau) [inline permuta/__init__.py:17]: This should be removed I think

File: permuta/misc/math.py [hunk 11]
```diff
@@ -10,3 +14,158 @@ def is_prime(n: int) -> bool:
...
+                    else:
+                        new_func[idx, n_idx] = self[pos]
+
+        return Matrix(len(self), elements=new_func)
+
+    def __hash__(self) -> int:
+        return hash(str(self))
+
+
+class LoTriMatrix(Matrix):
```
Comment (enadeau) [inline permuta/misc/math.py:163]: I don't think we need that class

File: permuta/__init__.py [hunk 12]
```diff
@@ -1,3 +1,4 @@
+from .misc.math import LoTriMatrix, Matrix
```
Comment (enadeau) [inline permuta/__init__.py:1]: Remove LoTriMatrix

**Review by enadeau — CHANGES_REQUESTED** [review]
I think we should remove the LoTriMatrix class since it was only used with the previous code. I've pointed the changes that needs to be made

**Review by enadeau — APPROVED** [review]

**Comment by enadeau** [comment]
Can you update de Changelog

**Pull request merged**